RFC-0008-plan-checker.md

```diff
@@ -0,0 +1,137 @@
+
+If a plan is generated that can't be executed in the cluster, the query will fail with an error message during the query execution
+phase.  This is a poor user experience, as the user has to wait for the query to fail before they can take corrective action, and
+this might have occurred after lengthy queueing.  Additionally, this failure would occur after worker resources have already been
+allocated to the query, which is wasteful.  This RFC proposes a plan checker that can be run before the query is executed to ensure
+a quick validation of the plan.
+
+### [Optional] Goals
+
+* Provide a mechanism to validate Presto to Velox plan conversion during planning phase
+* Add an SPI to allow custom validators to be added to suit individual business needs
+
+### [Optional] Non-goals
+
+* Ensure all Velox plans are executable in a Velox cluster--many checks are done at runtime and may not be caught by the plan checker
+
+## Proposed Implementation
+
+The Presto runtime centralizes plan validation logic into the `PlanChecker` class.  There exist three phases to this class: 
+
+* `validateIntermediatePlan`
+* `validateFinalPlan`
+* `validateFragmentedPlan`
+
+It is generally useful to allow this class to be configured with an SPI that allows for custom plan validation logic to be added.
+For example, a business may decide that a certain type is not allowed, or add a check to ensure that plans that are overly
+complicated are killed.
+
+An SPI will be added that will add more checks to the `PlanChecker` class.  The SPI will contain a field indicating which phase
+of the plan checker to be added, and a validator that will be run during that phase.  The SPI will be added to the `PlanChecker`
+class, and the `PlanChecker` class will be modified to run the SPIs during the appropriate phase.
+
+A new endpoint to the Presto sidecar will be added that will attempt to convert a Presto plan to a Velox plan.  In the
+`presto-native-plugin` module, a new implementation of the SPI will be added which will add a check to the `PlanChecker` class
+which will call the sidecar to attempt to convert the plan.  If the conversion fails, the plan checker will fail the plan.  It will
+be added at the `validateFragmentedPlan` phase--this is because it's not until plan fragmentation occurs that we know which portions
+of the plan will be executed in the coordinator, and which will be executed in the workers.
+
+The code for the plan checker will run `PrestoToVeloxQueryPlan`, which is used by the workers to convert the Presto plan fragment
+to a Velox plan fragment.  If the conversion fails, the plan checker will fail the plan, returning with it the reason for the failure.
+
+#### REST endpoint
+
+> Endpoint: /v1/plan
+>
+> HTTP verb: POST
+>
+> Request body: serialized plan fragment
+>
+> Response body: JSON response indicating success, or if failure, the reason for the failure
+
+#### Additional information
+
+1. What modules are involved
+   2. `presto-native-sidecar`
```

Comment:
this module doesn't exist yet?

Reply:
This will be a new module that is also being developed for other purposes such as function validation.  I will make that note in this RFC.